File: cpp/Maximum_Subarray.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_so_far = INT32_MIN;
        int max_end_here = 0;

        for(int num : nums) {
            max_end_here += num;
            if(max_end_here > max_so_far) max_so_far = max_end_here;
            if(max_end_here < 0) max_end_here = 0;
        }

        return max_so_far;
    }
};

int main(){
    // vector<int> inp = {-2,1,-3,4,-1,2,1,-5,4};
    // vector<int> inp = {1};
    vector<int> inp = {-5,-4,-1,-7,-8};

    Solution sol;
    cout <<sol.maxSubArray(inp) <<endl;

    return 0;
}

File: ruby/maximum_subarray.rb
class Integer
  N_BYTES = [42].pack('i').size
  N_BITS = N_BYTES * 16
  MAX = 2**(N_BITS - 2) - 1
  MIN = -MAX - 1
end

# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
  sum = 0
  max_sum = Integer::MIN

  nums.each do |num|
    sum += num
    max_sum = [sum, max_sum].max
    sum = 0 if sum.negative?
  end

  max_sum
end
